pkg/providers/vsphere/validator.go

```diff
@@ -0,0 +1,292 @@
+		return err
+	}
+
+	if err := p.govc.ValidateVCenterSetup(ctx, vsphereClusterSpec.datacenterConfig, &p.selfSigned); err != nil { // TODO: remove side effects from this implementation or directly move it to set defaults (try not to pass a pointer to bool to update its value, it's very difficult to follow)
+		return fmt.Errorf("error validating vCenter setup: %v", err)
+	}
+
+	for _, config := range vsphereClusterSpec.machineConfigsLookup {
+		err := p.govc.ValidateVCenterSetupMachineConfig(ctx, vsphereClusterSpec.datacenterConfig, config, &p.selfSigned) // TODO: remove side effects from this implementation or directly move it to set defaults (pointer to bool is not needed)
+		if err != nil {
+			return fmt.Errorf("error validating vCenter setup for VSphereMachineConfig %v: %v", config.Name, err)
+		}
+	}
+
+	if controlPlaneMachineConfig.Spec.OSFamily != workerNodeGroupMachineConfig.Spec.OSFamily {
+		return errors.New("control plane and worker nodes must have the same osFamily specified")
+	}
+
+	if etcdMachineConfig != nil && controlPlaneMachineConfig.Spec.OSFamily != etcdMachineConfig.Spec.OSFamily {
+		return errors.New("control plane and etcd machines must have the same osFamily specified")
+	}
+
+	if err := p.validateSSHUsername(controlPlaneMachineConfig); err == nil {
+		if err = p.validateSSHUsername(workerNodeGroupMachineConfig); err != nil {
+			return fmt.Errorf("error validating SSHUsername for worker node VSphereMachineConfig %v: %v", workerNodeGroupMachineConfig.Name, err)
+		}
+		if etcdMachineConfig != nil {
+			if err = p.validateSSHUsername(etcdMachineConfig); err != nil {
+				return fmt.Errorf("error validating SSHUsername for etcd VSphereMachineConfig %v: %v", etcdMachineConfig.Name, err)
+			}
+		}
+	} else {
+		return fmt.Errorf("error validating SSHUsername for control plane VSphereMachineConfig %v: %v", controlPlaneMachineConfig.Name, err)
+	}
+
+	for _, machineConfig := range vsphereClusterSpec.machineConfigsLookup {
+		if machineConfig.Namespace != vsphereClusterSpec.Cluster.Namespace {
+			return errors.New("VSphereMachineConfig and Cluster objects must have the same namespace specified")
+		}
+	}
+
+	if vsphereClusterSpec.datacenterConfig.Namespace != vsphereClusterSpec.Cluster.Namespace {
+		return errors.New("VSphereDatacenterConfig and Cluster objects must have the same namespace specified")
+	}
+
+	if err := p.validateTemplate(ctx, vsphereClusterSpec, controlPlaneMachineConfig); err != nil {
+		logger.V(1).Info("Control plane template validation failed.")
+		return err
+	}
+
+	// TODO: not sure if this makes any sense since we later validate than the two templates are the same?
+	if controlPlaneMachineConfig.Spec.Template != workerNodeGroupMachineConfig.Spec.Template {
+		if err := p.validateTemplate(ctx, vsphereClusterSpec, workerNodeGroupMachineConfig); err != nil {
+			logger.V(1).Info("Workload template validation failed.")
+			return err
+		}
+	}
```

Comment:
yep it doesn't make sense since both cp and worker templates must be the same. We prolly don't need this

Reply:
Removed